jumanji/environments/packing/job_shop/improvement/get_actions.py

```diff
@@ -0,0 +1,296 @@
+
+    # Check if the operations are adjacent in time (finish of one == start of the next)
+    is_adjacent = jnp.isclose(est[safe_senders] + ops_durations[safe_senders], est[safe_receivers])
+
+    # A pair is part of a critical block if:
+    # (1) both ops are valid (not padded),
+    # (2) both are on the critical path,
+    # (3) they are adjacent in time.
+    critical_blocks_mask = jnp.logical_and(
+        valid_pairs, jnp.logical_and(is_critical_pair, is_adjacent)
+    )  # shape (max_num_edges,)
+    senders = jnp.where(critical_blocks_mask, senders, -1)
+    receivers = jnp.where(critical_blocks_mask, receivers, -1)
+
+    # Sort pairs by earliest start time, with +inf for non-critical pairs
+    filtered_est = jnp.where(critical_blocks_mask, est[senders], jnp.inf)  # shape (max_num_edges,)
+    sorted_est = jnp.argsort(filtered_est)
+    critical_block_pairs = jnp.stack(
+        [senders[sorted_est], receivers[sorted_est]], axis=-1
+    )  # shape (max_num_edges, 2)
+
+    # === Initialize critical_block_info ===
+
+    # critical_block_info is a matrix of shape (num_ops_total, 8) containing:
+    # (is_on_critical_path, num_critical_block, is_left, is_right, left_neighbor, right_neighbor,
+    # left_end, right_end)
+
+    critical_block_info = jnp.zeros((num_ops_total, 8), dtype=jnp.int32)
+
+    # Set is_on_critical_path
+    critical_block_info = critical_block_info.at[:, CBFields.IS_ON_CRITICAL_PATH].set(critical_mask)
+    # Each operation initially defines its own critical block,
+    # set up num_critical_block and left_end accordingly (updated during scan)
+    critical_block_info = critical_block_info.at[:, CBFields.BLOCK_ID].set(
+        jnp.arange(num_ops_total)
+    )
+    critical_block_info = critical_block_info.at[:, CBFields.LEFT_END].set(
+        jnp.arange(num_ops_total)
+    )
+    # All ops are initially marked as right ends (updated during scan)
+    critical_block_info = critical_block_info.at[:, CBFields.IS_RIGHT].set(1)
+    # Initialize neighbors and right_end to -1
+    critical_block_info = critical_block_info.at[:, CBFields.LEFT_NEIGHBOR].set(-1)
+    critical_block_info = critical_block_info.at[:, CBFields.RIGHT_NEIGHBOR].set(-1)
+    critical_block_info = critical_block_info.at[:, CBFields.RIGHT_END].set(-1)
+
+    # Initialize right_end_array to -1
+    # Right end array is used to
+    right_end_array = jnp.full(num_ops_total, -1, dtype=jnp.int32)
+
+    def cond_fun(loop_state: Tuple[jnp.int32, Tuple[chex.Array, chex.Array]]) -> jnp.bool_:
+        """Condition function for the while loop.
+        The iteration stops when we have processed all the valid critical block pairs (i.e.
+        while start_idx and end_idx are both not -1).
+        """
+        i, _ = loop_state
+        start_idx, end_idx = critical_block_pairs[i]
+        return jnp.logical_and(i < num_ops_total, jnp.logical_and(start_idx != -1, end_idx != -1))
+
+    def body_fun(
```

Comment:
I don't know if this comment should be on this function or the while loop, but can you quickly explain the core of the algorithm you came up with here ? Mainly what you explained orally, how the data flows in the while loop